Source/WebCore/platform/network/curl/CurlMultipartHandle.cpp

```diff
@@ -32,326 +32,278 @@
-        auto length = matchedLength(content + i);
-        if (length == m_boundary.length()) {
-            boundaryStartPosition = i;
-            return true;
+    // Check the Close Delimiter
+    if (contentEndPtr - matchedEndPtr < 2) {
+        if (!m_didCompleteMessage) {
+            // Not enough data to check the Close Delimiter.
+            return result;
         }
-
-        if (length)
-            lastPartialMatchPosition = i;
+    } else if (!memcmp(matchedEndPtr, "--", 2)) {
+        result.hasBoundary = true;
+        result.hasCloseDelimiter = true;
+        result.processed = matchedEndPtr + 2 - contentStartPtr;
+        return result;
     }
 
-    return false;
-}
-
-size_t CurlMultipartHandle::matchedLength(const uint8_t* data)
-{
-    auto length = m_boundary.length();
-    for (size_t i = 0; i < length; ++i) {
-        if (data[i] != m_boundary[i])
-            return i;
+    // Skip transport-padding
+    for (; matchedEndPtr < contentEndPtr && isTabOrSpace(*matchedEndPtr); ++matchedEndPtr) { }
+
+    // There should be a \r and a \n but it seems that's not the case.
+    // So we'll check for a simple \n. Not really RFC compatible but servers do tricky things.
+    if (contentEndPtr - matchedEndPtr >= 2  && !memcmp(matchedEndPtr, "\r\n", 2))
+        matchedEndPtr += 2;
+    else if (contentEndPtr - matchedEndPtr >= 1 && !memcmp(matchedEndPtr, "\n", 1))
+        matchedEndPtr++;
+    else if (matchedEndPtr >= contentEndPtr) {
+        // Not enough data to check the boundary
+        return result;
+    } else {
+        result.isSyntaxError = true;
+        return result;
     }
 
-    return length;
+    result.hasBoundary = true;
+    result.processed = matchedEndPtr - contentStartPtr;
+    return result;
 }
 
-bool CurlMultipartHandle::parseHeadersIfPossible()
+std::optional<bool> CurlMultipartHandle::parseHeadersIfPossible()
 {
-    auto contentLength = m_buffer.size();
+    static const auto maxHeaderSize = 300 * 1024;
 
-    if (!contentLength)
-        return false;
-
-    auto* content = m_buffer.data();
+    auto contentLength = m_buffer.size();
+    const auto contentStartPtr = m_buffer.data();
 
     // Check if we have the header closing strings.
-    if (!memmem(content, contentLength, "\r\n\r\n", 4)) {
-        // Some servers closes the headers with only \n-s.
-        if (!memmem(content, contentLength, "\n\n", 2)) {
-            // Don't have the header closing string. Wait for more data.
-            return false;
-        }
+    const uint8_t* end = nullptr;
+    if ((end = static_cast<const uint8_t*>(memmem(contentStartPtr, contentLength, "\r\n\r\n", 4))))
+        end += 4;
+    else if ((end = static_cast<const uint8_t*>(memmem(contentStartPtr, contentLength, "\n\n", 2))))
+        end += 2;
+    else {
+        // Don't have the header closing string. Wait for more data.
+        return false;
     }
 
+    if (end - contentStartPtr > maxHeaderSize)
+        return std::nullopt;
```

Comment:
You should use enum or `std::pair<bool, bool>` for the return type rather than `std::nullopt`.

Reply:
I changed to enum. Fixed.